cmd/orderimports/orderimports.go

```diff
@@ -0,0 +1,210 @@
+		a.failed = true
+		return
+	}
+
+	for _, p := range fs {
+		// returns first error, but a.handleError deals with it
+		files := a.filterFiles(p.Files)
+		for _, file := range files {
+			pathToFile := a.fset.File(file.Pos()).Name()
+
+			if len(file.Imports) <= 1 {
+				continue
+			}
+			var originalImports, stdlibImports, localImports, k8sImports, externalImports []string
+
+			for i, imp := range file.Imports {
+				importPath := strings.Replace(imp.Path.Value, "\"", "", -1)
+				parts := strings.Split(importPath, "/")
+
+				if i != 0 && a.lineAt(imp.Pos()) > 1+a.lineAt(file.Imports[i-1].End()) {
+					originalImports = append(originalImports, "")
+				}
+				originalImports = append(originalImports, importPath)
+
+				if !strings.Contains(parts[0], ".") {
+					// standard library imports are first
+					stdlibImports = append(stdlibImports, importPath)
+				} else if strings.HasPrefix(importPath, "k8s.io/kubernetes") {
+					// local imports are second
+					localImports = append(localImports, importPath)
+				} else if strings.Contains(parts[0], "k8s.io") {
+					// other *.k8s.io imports are third
+					k8sImports = append(k8sImports, importPath)
+				} else {
+					// external repositories are fourth
+					externalImports = append(externalImports, importPath)
+				}
+			}
+
+			orderImports := []string{}
+			for _, imps := range [][]string{
+				stdlibImports,
+				localImports,
+				k8sImports,
+				externalImports,
+			} {
+				sort.Strings(imps)
+				orderImports = append(orderImports, imps...)
+				if len(imps) > 0 {
+					orderImports = append(orderImports, "")
+				}
+			}
+			if orderImports[len(orderImports)-1] == "" {
+				orderImports = orderImports[:len(orderImports)-1]
+			}
+
+			if diff := cmp.Diff(strings.Join(originalImports, "\n"), strings.Join(orderImports, "\n")); diff != "" {
+				a.failed = true
+				fmt.Printf("%s (-got +want):\n%s", pathToFile, diff)
+			}
+		}
+	}
+}
+
+func (a *analyzer) lineAt(pos token.Pos) int {
+	return a.fset.PositionFor(pos, false).Line
+}
+
+func (a *analyzer) filterFiles(fs map[string]*ast.File) []*ast.File {
+	var files []*ast.File
+	for fileName, f := range fs {
+		if a.ignoreFile.MatchString(fileName) {
+			continue
+		}
+		files = append(files, f)
+	}
+	return files
+}
+
+type collector struct {
+	dirs        []string
+	includePath *regexp.Regexp
+}
+
+// handlePath walks the filesystem recursively, collecting directories,
+// ignoring some unneeded directories (hidden/vendored) that are handled
+// specially later.
+func (c *collector) handlePath(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		return err
+	}
+	if info.IsDir() {
+		// Ignore hidden directories (.git, .cache, etc)
+		if len(path) > 1 && path[0] == '.' ||
+			// Staging code is symlinked from vendor/k8s.io, and uses import
+			// paths as if it were inside of vendor/. It fails typechecking
+			// inside of staging/, but works when typechecked as part of vendor/.
+			path == "staging" ||
```

Comment:
if we are going to test /staging eventually with this tool, i think this is problematic, because staging also contains original source code that will be skipped?

@xinydev some context, /staging is used to publish to external repositories like https://git.k8s.io/kubectl

Reply:
Done.

These folders(`staging`,`vendor`) are no longer skipped. will consider whether to skip these folders when we are ready to open this to the rest of k/k

Reply:
We should skip vendor because its third party code. Staging is our code and we may want to lint it eventually.

My comment about vendor was related to the symlinks which was mentioned in the code.

Reply:
Sorry, I misunderstood, just fixed